### PR TITLE
Add 6328's implementation of PNP distance for Trig Solving to PhotonPoseEstimator

### PR DESCRIPTION
**Pull request by JuliusZhou124**
https://discord.com/channels/725836368059826228/725846784131203222/1334309604946874460

https://www.chiefdelphi.com/t/frc-6328-mechanical-advantage-2025-build-thread/477314/85

Helps with ambiguous single tag estimates and produces more stability.

### CODE REVIEW

**Comment by mcm001** [comment]
Sweet. I'd like to see some tests as well if we can.

**Comment by jwbonner** [comment]
I would recommend taking a look at [this post](https://www.chiefdelphi.com/t/frc-6328-mechanical-advantage-2025-build-thread/477314/114?u=jonahb55) and the reply. The original version of this logic that we released (which seems to be implemented here) doesn't work properly when ty is nonzero.

**Comment by JuliusZhou124** [comment]
> I would recommend taking a look at [this post](https://www.chiefdelphi.com/t/frc-6328-mechanical-advantage-2025-build-thread/477314/114?u=jonahb55) and the reply. The original version of this logic that we released (which seems to be implemented here) doesn't work properly when ty is nonzero.

Thanks for the catch! Missed that in the thread. 😅

**Comment by mcm001** [comment]
Fya, we will still need to see tests for all the edge cases (camera straight, pitched/yawed/rolled, compound angles) before merging. I think our simulation code should be able to get you the fake targets you need, and you can just copy paste some of our existing pose estimator tests.

**Comment by JuliusZhou124** [comment]
> I think our simulation code should be able to get you the fake targets you need, and you can just copy paste some of our existing pose estimator tests.

Is there any fast way to get fake targets from simulation? Brute force copying seems rather tedious.

**Comment by mcm001** [comment]
Yeah. I think you should be able to use our vision system simulation to generate the target data?

**Comment by JuliusZhou124** [comment]
Looks like the poses begin to deviate as the targets approach the corners of the camera feed. 
image

Might be an implementation bug? Either way it's 10+ cm off in both x and y when simulating with perfect cameras.

**Comment by JuliusZhou124** [comment]
Struggles with rolled cameras as well: 
image

**Comment by mcm001** [comment]
Broadly Lgtm. Delegating to @spacey-sooty :)

File: photon-lib/src/test/java/org/photonvision/PhotonPoseEstimatorTest.java [hunk 1]
```diff
@@ -489,6 +491,125 @@ void closestToLastPose() {
         assertEquals(1, pose.getZ(), .01);
     }
 
+    @Test
+    void pnpDistanceTrigSolve() {
+        PhotonCameraInjector cameraOne = new PhotonCameraInjector();
+
+        // With Roll, Pitch, Yaw
```
Comment (mcm001) [inline photon-lib/src/test/java/org/photonvision/PhotonPoseEstimatorTest.java:498]: Please document where these magic results came from, or consider using a vision system simulation to generate them dynamically here in the test.

Comment (JuliusZhou124) [reply photon-lib/src/test/java/org/photonvision/PhotonPoseEstimatorTest.java:498]: The sim can dynamically generate the data? 🫠

I just simmed a perfect camera in my robot code with the lowest ambiguity strategy with the apriltag map and manually copied over target data from logs 💀

Comment (mcm001) [reply photon-lib/src/test/java/org/photonvision/PhotonPoseEstimatorTest.java:498]: Using the sim would me more maintainable for sure. You could try using PhotonCameraSim directly?

https://github.com/PhotonVision/photonvision/blob/main/photon-lib/src/main/java/org/photonvision/simulation/PhotonCameraSim.java#L370

File: photon-lib/src/test/java/org/photonvision/PhotonPoseEstimatorTest.java [hunk 2]
```diff
@@ -489,6 +493,71 @@ void closestToLastPose() {
...
+                        PoseStrategy.PNP_DISTANCE_TRIG_SOLVE,
+                        new Transform3d(
+                                -Units.inchesToMeters(12),
+                                -Units.inchesToMeters(11),
+                                3,
+                                new Rotation3d(
+                                        Units.degreesToRadians(37),
+                                        Units.degreesToRadians(6),
+                                        Units.degreesToRadians(60))));
+
```
Comment (mcm001) [inline photon-lib/src/test/java/org/photonvision/PhotonPoseEstimatorTest.java:520]: Just to be explicit, we could consider pulling this transform out into a variable?

**Review by mcm001 — APPROVED** [review]

File: photon-lib/src/test/java/org/photonvision/PhotonPoseEstimatorTest.java (hunk 2 above)
Comment (JuliusZhou124) [reply photon-lib/src/test/java/org/photonvision/PhotonPoseEstimatorTest.java]: Yup sounds good

**Review by mcm001 — APPROVED** [review]

**Comment by mcm001** [comment]
Looks like two new flakes:
- https://github.com/PhotonVision/photonvision/actions/runs/13314440628/job/37184905664?pr=1767#step:5:4580 
```
 Starting process 'command '/usr/bin/clang++''. Working directory: /Users/runner/work/photonvision/photonvision/photon-targeting/build/objs/photontargetingTest/osxuniversal/release/photontargetingCpp Command: /usr/bin/clang++ @/Users/runner/work/photonvision/photonvision/photon-targeting/build/tmp/compilePhotontargetingTestOsxuniversalReleaseGoogleTestExePhotontargetingCpp/options.txt /Users/runner/work/photonvision/photonvision/photon-targeting/src/main/native/cpp/photon/targeting/MultiTargetPNPResult.cpp -o /Users/runner/work/photonvision/photonvision/photon-targeting/build/objs/photontargetingTest/osxuniversal/release/photontargetingCpp/at79tvvs4oj7e6ruemd6xlbm5/MultiTargetPNPResult.o
Successfully started process 'command '/usr/bin/clang++''
libc++abi: terminating due to uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument
```
- https://github.com/PhotonVision/photonvision/actions/runs/13314440628/job/37184904590?pr=1767#step:5:7240
```
 [ RUN      ] VisionSystemSimTest.TestMultipleTargets
D:\a\photonvision\photonvision\photon-lib\src\test\native\cpp\VisionSystemSimTest.cpp(401): error: Expected equality of these values:
  static_cast<size_t>(11)
    Which is: 11
  tgtList.size()
    Which is: 10

[  FAILED  ] VisionSystemSimTest.TestMultipleTargets (12 ms)
```

**Comment by mcm001** [comment]
@JuliusZhou124 as follow on work, we should update our docs page and stuff. And you should go brag about this in chief delphi!
